### DISCUSSION TITLE
Calculating the amount of observations inside percentiles is not the same after using Mixture Density Networks

### DISCUSSION THREAD

**Discussion by maby200** [neuralhydrology/neuralhydrology#172] [Q&A] [answered]
I have been using a MDN (Gaussian Mixtures) for obtaining uncertainties. What I want to know is the amount of observations `y` inside the percentiles 2.5 and 97.5 (known as p-factor for some [researchers](https://www.sciencedirect.com/science/article/pii/S0022169415001985?ref=pdf_download&fr=RR-2&rr=8636a5dc0d904ffb) ). For that, used the function [uncertainty_plot](https://github.com/neuralhydrology/neuralhydrology/blob/master/neuralhydrology/evaluation/plots.py) as a guide to calculate it.

**But something I noticed is that the resulting value changes each time I restart the kernel.** 
#### What I would like to know is if there is a way to make it a fixed result (afaik it should be) or maybe there is a way to explain this variation.

How I am loading my results before calculating everything:
```python
run_dir = Path("runs/test_run_1309_182624")
eval_run(run_dir=run_dir, period="test")
with open(run_dir / "test" / "model_epoch039" / "test_results.p", "rb") as fp:
    results = pickle.load(fp)
```

Here is the code for the `p-factor`:
```python
def p_factor(y: np.ndarray, y_hat: np.ndarray, basin_name:str):
  # extract the last column of y
  y_long = y[:,-1].flatten()
  # reshape y_hat to match the shape of y_long
  y_hat_long = y_hat[:,-1,:].reshape(y_long.shape[0],-1)
  # generate x_bnd using the length of y
  x_bnd = np.arange(0,len(y))

  # lower and upper bounds for the percentile
  lb = round(50 - (0.95 * 100) / 2)
  ub = round(50 + (0.95 * 100) / 2)
  y_lb = np.percentile(y_hat_long[x_bnd, :], lb, axis=-1).flatten()
  y_ub = np.percentile(y_hat_long[x_bnd, :], ub, axis=-1).flatten()
  
  p_factor = y_long[ (y_long[x_bnd] <= y_ub[x_bnd]) & (y_long[x_bnd] >= y_lb[x_bnd] )]
  p_factor = np.divide(p_factor.size,x_bnd.size)
  
  return f'p-factor in {basin_name}: {p_factor:.4f}'
``` 

I tried a `random.seed(1000)` but not fixed this.

Here is a sample of the results that I obtain after calculation of the p-factor:
|Basin|1st result|2nd result|3rd result|
|:----:|:-----:|:----:|:---:|
basin_1 	|	.7812 |	.7810|	.7749|	.7818
basin_2 	|	.7542 |	.7697|	.7247|	.7645
basin_3 	|	.7220 |	.7169|	.7515|	.7143

**Comment by gauchm** [comment, marked as answer]
Sampling during inference draws randomness from torch.random (maybe also numpy.random, not sure), so you need to set other seeds. This should work (probably this is even more than what's really necessary):

```
random.seed(0)
np.random.seed(0)
torch.cuda.manual_seed(0)
torch.manual_seed(0)
```

**Reply by maby200** [reply]
Thank you, this fixed the values each time kernel was restarted.
